File: Sean的leetcode/FB/3Sum.py
###
# Time Complexity: O(n^2)
# Space Complexity: O(1)
###

class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        ln = len(nums)
        if ln < 3:
            return []
        re = []
        nums.sort()
        i = 0
        while i < ln-2:
            l = i + 1
            r = ln - 1
            while l < r:
                s = nums[i] + nums[l] + nums[r]
                if s < 0:
                    l += 1
                elif s > 0:
                    r -= 1
                else:
                    re += [nums[i], nums[l], nums[r]],
                    l += 1
                    r -= 1
                    while l < r and nums[l] == nums[l-1]:
                        l += 1
                    while l < r and nums[r] == nums[r+1]:
                        r -= 1
            i += 1
            while i < ln-2 and nums[i] == nums[i-1]:
                i += 1
        return re

File: Sean的leetcode/Leetcode/3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int> >res;
        vector<int>one_sol;
        int n = nums.size();
        int right = nums.size() - 1;

        sort(nums.begin(), nums.end());

        for (int i = 0 ; i < n - 2;)
        {
            one_sol.push_back(nums[i]);
            for (int left = i + 1, right = n - 1; left < right;)
            {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum > 0)
                    right--;
                else if (sum < 0)
                    left++;
                else
                {
                    one_sol.push_back(nums[left]);
                    one_sol.push_back(nums[right]);
                    res.push_back(one_sol);
                    one_sol.pop_back();
                    one_sol.pop_back();

                    left++, right--;
                    while (left < right && nums[left] == nums[left-1])
                        left++;

                    while (left < right && nums[right] == nums[right+1])
                        right--;
                }
            }
            one_sol.pop_back();
            i++;
            while (i < n - 2  && nums[i-1] == nums[i])
                i++;
        }
        return res;
    }
};